File: repowire/daemon/scheduler.py
"""Scheduled check-in dispatcher.

A single background task sleeps until the next-due schedule's `fire_at`, then
fires it. One-shot schedules are removed after firing; recurring cron
schedules are advanced to their next fire time. No polling: the wake event is
only set when the schedule set changes (create/delete) or when something fires.
Past-due schedules at startup fire immediately.

Delivery reuses the existing peer-registry primitives:
  - kind="notify" -> PeerRegistry.notify (fire-and-forget)
  - kind="ask"    -> PeerRegistry.deliver_ask (registers in AskTracker first)

On delivery failure (peer missing or no live WS), one-shot schedules are
dropped and recurring schedules advance to their next cron fire. We log loudly
so misfires are visible.
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime, timezone
from typing import TYPE_CHECKING
from uuid import uuid4

from repowire.daemon.schedule_store import Schedule, ScheduleStoreProtocol
from repowire.daemon.websocket_transport import TransportError

if TYPE_CHECKING:
    from repowire.daemon.ask_tracker import AskTracker
    from repowire.daemon.peer_registry import PeerRegistry

logger = logging.getLogger(__name__)

# Cap a single sleep so a long horizon doesn't outlive a clock jump unnoticed.
_MAX_SLEEP_SECONDS = 3600.0


class Scheduler:
    """Drives one-shot scheduled check-ins."""

    def __init__(
        self,
        store: ScheduleStoreProtocol,
        peer_registry: PeerRegistry,
        ask_tracker: AskTracker,
    ) -> None:
        self._store = store
        self._peer_registry = peer_registry
        self._ask_tracker = ask_tracker
        self._wake = asyncio.Event()
        self._task: asyncio.Task[None] | None = None
        self._stopped = False

    async def start(self) -> None:
        if self._task is not None:
            return
        self._stopped = False
        self._task = asyncio.create_task(self._run(), name="scheduler")

    async def stop(self) -> None:
        self._stopped = True
        self._wake.set()
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except (asyncio.CancelledError, Exception):
                pass
            self._task = None
        self._store.persist()

    def notify_changed(self) -> None:
        """Signal the loop that the schedule set changed (create/delete)."""
        self._wake.set()

    async def _run(self) -> None:
        while not self._stopped:
            self._wake.clear()
            nxt = self._store.next_due()
            if nxt is None:
                await self._wake.wait()
                continue
            now = datetime.now(timezone.utc)
            delay = (nxt.fire_at_dt() - now).total_seconds()
            if delay > 0:
                try:
                    await asyncio.wait_for(
                        self._wake.wait(),
                        timeout=min(delay, _MAX_SLEEP_SECONDS),
                    )
                    # Woken by a change — reloop and recompute next due.
                    continue
                except asyncio.TimeoutError:
                    pass
            # Re-check that this schedule still exists (delete may have raced).
            current = self._store.get(nxt.schedule_id)
            if current is None:
                continue
            await self._fire(current)

    async def _fire(self, sched: Schedule) -> None:
        try:
            if sched.kind == "ask":
                cid = f"ask-{uuid4().hex[:8]}"
                await self._ask_tracker.register(
                    from_peer_id=sched.from_peer,
                    from_peer_name=sched.from_peer,
                    to_peer_id=sched.to_peer,
                    to_peer_name=sched.to_peer,
                    text=sched.text,
                    correlation_id=cid,
                )
                try:
                    await self._peer_registry.deliver_ask(
                        from_peer=sched.from_peer,
                        to_peer=sched.to_peer,
                        text=sched.text,
                        correlation_id=cid,
                        circle=sched.circle,
                    )
                except (ValueError, TransportError) as e:
                    await self._ask_tracker.close(cid, reason="send_failed")
                    logger.warning(
                        "Scheduled ask %s failed to deliver: %s",
                        sched.schedule_id, e,
                    )
            else:
                try:
                    await self._peer_registry.notify(
                        from_peer=sched.from_peer,
                        to_peer=sched.to_peer,
                        text=sched.text,
                        circle=sched.circle,
                    )
                except (ValueError, TransportError) as e:
                    logger.warning(
                        "Scheduled notify %s failed to deliver: %s",
                        sched.schedule_id, e,
                    )
            logger.info(
                "Fired schedule %s: %s -> %s (%s)",
                sched.schedule_id, sched.from_peer, sched.to_peer, sched.kind,
            )
        except Exception as e:
            # Catch-all so a misbehaving delivery (anything beyond the inner
            # ValueError/TransportError handlers above) doesn't escape and
            # kill the scheduler task. The loop continues to the next
            # schedule; this one still gets dropped via the finally block.
            logger.exception(
                "Scheduled %s %s: unexpected error during fire (%s); dropping",
                sched.kind, sched.schedule_id, e,
            )
        finally:
            if sched.cron:
                try:
                    self._store.reschedule_next(sched.schedule_id)
                except ValueError as e:
                    logger.warning(
                        "Recurring schedule %s has invalid cron after fire (%s); dropping",
                        sched.schedule_id, e,
                    )
                    self._store.delete(sched.schedule_id)
            else:
                self._store.delete(sched.schedule_id)
            self._store.persist()

File: daemon-go/service/scheduler.go
package service

// Scheduled check-in dispatcher. Port of repowire/daemon/scheduler.py.
//
// A single goroutine sleeps until the next-due schedule's fire_at, then fires
// it. One-shot schedules are deleted after firing; recurring cron schedules
// advance to their next fire time. No polling: the wake channel is only
// signalled when the schedule set changes (create/delete via Wake) or after
// something fires. Past-due schedules fire immediately.
//
// Delivery reuses PeerDelivery so scheduled ask/notify traffic follows the same
// application delivery path as route-driven traffic. On delivery failure (peer
// missing or no live transport) one-shot schedules are still dropped and
// recurring ones advance — we log loudly so misfires are visible (fail loud over
// silent degrade).

import (
	"context"
	"log"
	"time"

	"github.com/repowire/repowire/daemon-go/state"
)

// maxSchedulerSleep caps a single sleep so a long horizon doesn't outlive a
// clock jump unnoticed. Mirrors scheduler._MAX_SLEEP_SECONDS.
const maxSchedulerSleep = time.Hour

// schedulerStore is the schedule data layer the loop drives. *state.Store
// satisfies it.
type schedulerStore interface {
	NextDueSchedule(ctx context.Context) (*state.Schedule, error)
	GetSchedule(ctx context.Context, scheduleID string) (*state.Schedule, error)
	RescheduleNext(ctx context.Context, scheduleID string, nextFireAt time.Time) (bool, error)
	DeleteSchedule(ctx context.Context, scheduleID string) (*state.Schedule, error)
}

// schedulerDelivery is the delivery seam (notify + scheduled-ask). *PeerDelivery
// satisfies it.
type schedulerDelivery interface {
	Notify(ctx context.Context, params NotifyParams) (NotifyResult, error)
	OpenScheduledAsk(ctx context.Context, fromPeer, toPeer, text string, circle *string, replyDelivery string) (string, error)
}

// Scheduler drives scheduled check-ins off a deadline-driven sleep + wake
// channel (never a poll timer).
type Scheduler struct {
	store    schedulerStore
	delivery schedulerDelivery

	wake   chan struct{}
	stop   chan struct{}
	doneCh chan struct{}
}

// NewScheduler wires the dispatcher. delivery is the PeerDelivery the routes
// already build; store is the schedules data layer.
func NewScheduler(store schedulerStore, delivery schedulerDelivery) *Scheduler {
	return &Scheduler{
		store:    store,
		delivery: delivery,
		// Buffered so a Wake() that races the loop's select isn't lost: the loop
		// always drains+rechecks next_due on each iteration.
		wake: make(chan struct{}, 1),
		stop: make(chan struct{}),
	}
}

// Wake signals the loop that the schedule set changed (create/delete). This is
// the notify_changed analogue: non-blocking, coalescing (a pending wake is not
// duplicated).
func (s *Scheduler) Wake() {
	select {
	case s.wake <- struct{}{}:
	default:
	}
}

// Start launches the firing goroutine. Idempotent: a second Start is a no-op
// while one is running.
func (s *Scheduler) Start(ctx context.Context) {
	if s.doneCh != nil {
		return
	}
	s.doneCh = make(chan struct{})
	go s.run(ctx)
}

// Stop signals the loop to exit and waits for the goroutine to finish.
func (s *Scheduler) Stop() {
	if s.doneCh == nil {
		return
	}
	close(s.stop)
	<-s.doneCh
	s.doneCh = nil
}

func (s *Scheduler) run(ctx context.Context) {
	defer close(s.doneCh)
	for {
		// Drain any pending wake before computing the next deadline so a change
		// signalled while we were firing is consumed here, not on the next loop.
		select {
		case <-s.wake:
		default:
		}

		nxt, err := s.store.NextDueSchedule(ctx)
		if err != nil {
			log.Printf("scheduler: next_due failed: %v; backing off", err)
			if s.sleep(ctx, maxSchedulerSleep) {
				return
			}
			continue
		}
		if nxt == nil {
			// Nothing scheduled: block until woken (or stopped).
			if s.waitForWake(ctx) {
				return
			}
			continue
		}

		fireAt, perr := parseScheduleFireAt(nxt.FireAt)
		if perr != nil {
			// A schedule with an unparseable fire_at can never fire; drop it loudly
			// rather than wedging the loop on it forever.
			log.Printf("scheduler: schedule %s has unparseable fire_at %q (%v); dropping", nxt.ScheduleID, nxt.FireAt, perr)
			_, _ = s.store.DeleteSchedule(ctx, nxt.ScheduleID)
			continue
		}

		delay := time.Until(fireAt)
		if delay > 0 {
			if delay > maxSchedulerSleep {
				delay = maxSchedulerSleep
			}
			woken, stopped := s.sleepUntil(ctx, delay)
			if stopped {
				return
			}
			if woken {
				// Schedule set changed (or sleep was capped): reloop, recompute.
				continue
			}
		}

		// Re-check the schedule still exists (a delete may have raced the sleep).
		current, err := s.store.GetSchedule(ctx, nxt.ScheduleID)
		if err != nil {
			log.Printf("scheduler: get %s failed: %v; skipping", nxt.ScheduleID, err)
			continue
		}
		if current == nil {
			continue
		}
		s.fire(ctx, current)
	}
}

// fire delivers a schedule then advances (cron) or removes (one-shot) it. Any
// delivery error is logged but never propagated — a misbehaving delivery must
// not kill the loop, and the schedule is still advanced/dropped in the finally
// equivalent below. Mirrors Scheduler._fire.
func (s *Scheduler) fire(ctx context.Context, sched *state.Schedule) {
	if sched.Kind == "ask" {
		if _, err := s.delivery.OpenScheduledAsk(ctx, sched.FromPeer, sched.ToPeer, sched.Text, sched.Circle, "push"); err != nil {
			log.Printf("scheduler: scheduled ask %s failed to deliver: %v", sched.ScheduleID, err)
		}
	} else {
		if _, err := s.delivery.Notify(ctx, NotifyParams{
			FromPeer: sched.FromPeer,
			ToPeer:   sched.ToPeer,
			Text:     sched.Text,
			Circle:   sched.Circle,
		}); err != nil {
			log.Printf("scheduler: scheduled notify %s failed to deliver: %v", sched.ScheduleID, err)
		}
	}
	log.Printf("scheduler: fired schedule %s: %s -> %s (%s)", sched.ScheduleID, sched.FromPeer, sched.ToPeer, sched.Kind)

	// Advance (cron) or drop (one-shot). Always runs regardless of delivery
	// outcome, mirroring the Python finally block.
	if sched.Cron != nil {
		next, err := NextFireAfter(*sched.Cron, time.Now().UTC())
		if err != nil {
			log.Printf("scheduler: recurring schedule %s has invalid cron after fire (%v); dropping", sched.ScheduleID, err)
			_, _ = s.store.DeleteSchedule(ctx, sched.ScheduleID)
			return
		}
		if _, err := s.store.RescheduleNext(ctx, sched.ScheduleID, next); err != nil {
			log.Printf("scheduler: reschedule %s failed (%v); dropping", sched.ScheduleID, err)
			_, _ = s.store.DeleteSchedule(ctx, sched.ScheduleID)
		}
		return
	}
	if _, err := s.store.DeleteSchedule(ctx, sched.ScheduleID); err != nil {
		log.Printf("scheduler: delete one-shot %s failed: %v", sched.ScheduleID, err)
	}
}

// sleepUntil sleeps for d, returning (woken, stopped). woken is true if a wake
// signal arrived first; stopped is true if Stop/ctx fired.
func (s *Scheduler) sleepUntil(ctx context.Context, d time.Duration) (woken, stopped bool) {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-t.C:
		return false, false
	case <-s.wake:
		return true, false
	case <-s.stop:
		return false, true
	case <-ctx.Done():
		return false, true
	}
}

// waitForWake blocks until a wake signal, Stop, or ctx cancellation. Returns
// true if the loop should exit.
func (s *Scheduler) waitForWake(ctx context.Context) (stopped bool) {
	select {
	case <-s.wake:
		return false
	case <-s.stop:
		return true
	case <-ctx.Done():
		return true
	}
}

// sleep blocks for d unless Stop/ctx fires first; returns true to exit.
func (s *Scheduler) sleep(ctx context.Context, d time.Duration) (stopped bool) {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-t.C:
		return false
	case <-s.stop:
		return true
	case <-ctx.Done():
		return true
	}
}

// parseScheduleFireAt parses the stored ISO-8601 fire_at.
func parseScheduleFireAt(raw string) (time.Time, error) {
	t, err := time.Parse(time.RFC3339Nano, raw)
	if err != nil {
		return time.Time{}, err
	}
	return t.UTC(), nil
}
